mail/api.py

```diff
@@ -322,18 +321,26 @@ def add_automatic_email(original_search, email_subject, email_body, sender_name,
         )
 
 
-@transaction.atomic
+@contextmanager
 def mark_emails_as_sent(automatic_email, emails):
     """
-    Mark users who have the given emails as sent
+    Context manager to mark users who have the given emails as sent after successful sending of email
 
     Args:
         automatic_email (AutomaticEmail): An instance of AutomaticEmail
         emails (iterable): An iterable of emails
     """
-    users = User.objects.filter(email__in=emails)
-    for user in users:
-        SentAutomaticEmail.objects.get_or_create(
-            user=user,
-            automatic_email=automatic_email,
-        )
+    with transaction.atomic():
+        user_ids = User.objects.filter(email__in=emails).exclude(
+            sentautomaticemail__automatic_email=automatic_email,
+        ).values_list('id', flat=True)
+
+        for user_id in user_ids:
+            SentAutomaticEmail.objects.create(
+                user_id=user_id,
+                automatic_email=automatic_email,
+                status=SentAutomaticEmail.PENDING,
+            )
+        SentAutomaticEmail.objects.filter(user_id__in=user_ids).select_for_update()
```

Comment:
Here you're locking all rows in `SentAutomaticEmail` for the list of users, I think you only want to lock the ones that  are in pending status for the particular `AutomaticEmail` you're trying to send here.